src/rgw/rgw_amqp_1.cc

```diff
@@ -0,0 +1,542 @@
+        // messages(max_queue),
+        queued(0),
+        dequeued(0),
+        cct(_cct),
+        container_runner(&Manager::run_container, this) {
+          // The hashmap has "max connections" as the initial number of buckets, 
+          // and allows for 10 collisions per bucket before rehash.
+          // This is to prevent rehashing so that iterators are not invalidated 
+          // when a new connection is added.
+          connections.max_load_factor(10.0);
+          // give the runner thread a name for easier debugging
+          // this name should not be too long or the assertion fail
+          const auto rc = ceph_pthread_setname(container_runner.native_handle(),
+              "amqp1.0 manager");
+          ceph_assert(rc==0);
+        }
+
+      // non copyable
+      Manager(const Manager&) = delete;
+      const Manager& operator=(const Manager&) = delete;
+
+      // stop the main thread
+      void stop() {
+        stopped = true;
+      }
+
+      // disconnect() is removed.
+
+      connection_ptr_t connect(const std::string& url, bool use_ssl,
+          boost::optional<const std::string&> ca_location) {
+        if(stopped) {
+          ldout(cct, 1) << "AMQP_1 connect: manager is stopped" << dendl;
+          return nullptr;
+        }
+        // TODO: max connections check
+        std::lock_guard<std::mutex> lock(connections_lock);
+
+        // TODO: find in the stale connectionlists
+
+        const auto conn = create_new_connection(container, url, cct, ca_location);
+
+        ceph_assert(conn);
+        ++connection_count;
+        ldout(cct, 10) << "AMQP_1 connect: new connection is created. Total"
+          "connections: " << connection_count << dendl;
+        return connections.emplace(url, conn).first->second;
+      }
+
+      int publish(connection_ptr_t& conn, const std::string& topic, const
+          std::string& message) {
+        if(stopped) {
+          return RGW_AMQP_1_STATUS_MANAGER_STOPPED;
+        }
+        if(!conn || !conn->is_ok()) {
+          return RGW_AMQP_1_STATUS_CONNECTION_CLOSED;
+        }
+        if(conn->messages.push(new message_wrapper_t(conn, topic, message, nullptr)))
+          {
+            ++queued;
+            return RGW_AMQP_1_STATUS_OK;
+          }
+        return RGW_AMQP_1_STATUS_QUEUE_FULL;
+      }
+
+      int publish_with_confirm(connection_ptr_t& conn, const std::string& topic,
+          const std::string& message, reply_callback_t cb) {
+        if(stopped) {
+          return RGW_AMQP_1_STATUS_MANAGER_STOPPED;
+        }
+        if(!conn || !conn->is_ok()) {
+          return RGW_AMQP_1_STATUS_CONNECTION_CLOSED;
+        }
+        if(conn->messages.push(new message_wrapper_t(conn, topic, message, cb)))
+          {
+            ++queued;
+            return RGW_AMQP_1_STATUS_OK;
+          }
+        return RGW_AMQP_1_STATUS_QUEUE_FULL;
+      }
+
+      // we have to close all active connection when delete Manager
```

Comment:
why?
why can't we just stop the container thread?

Reply:
Stopping the container thread would abandon all `works` in the work_queue. Proton container enable auto_stop by default, thus container::run() will return if there's no active connection. I'll fix this by directly stopping the container.

Reply:
trying to stop a connection from here is not thread safe.
so, the 2 options we have are:
* make it thread safe
* shutdown the container without closing the connections